test/integration/dnsrecord/dnsrecord_test.go

```diff
@@ -0,0 +1,294 @@
+				Name:      "dnsrecord",
+				Namespace: testName,
+			},
+			Data: map[string][]byte{
+				aws.AccessKeyID:     []byte(*accessKeyID),
+				aws.SecretAccessKey: []byte(*secretAccessKey),
+			},
+		}
+		cluster = &extensionsv1alpha1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+	)
+
+	BeforeSuite(func() {
+		repoRoot := filepath.Join("..", "..", "..")
+
+		// enable manager logs
+		logf.SetLogger(zap.New(zap.UseDevMode(true), zap.WriteTo(GinkgoWriter)))
+
+		log := logrus.New()
+		log.SetOutput(GinkgoWriter)
+		logger = logrus.NewEntry(log)
+
+		By("starting test environment")
+		testEnv = &envtest.Environment{
+			UseExistingCluster: pointer.BoolPtr(true),
+			CRDInstallOptions: envtest.CRDInstallOptions{
+				Paths: []string{
+					filepath.Join(repoRoot, "example", "20-crd-dnsrecord.yaml"),
+					filepath.Join(repoRoot, "example", "20-crd-cluster.yaml"),
+				},
+			},
+		}
+
+		cfg, err := testEnv.Start()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(cfg).ToNot(BeNil())
+
+		By("setting up manager")
+		mgr, err := manager.New(cfg, manager.Options{
+			MetricsBindAddress: "0",
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(extensionsv1alpha1.AddToScheme(mgr.GetScheme())).To(Succeed())
+		Expect(awsinstall.AddToScheme(mgr.GetScheme())).To(Succeed())
+
+		Expect(dnsrecordctrl.AddToManager(mgr)).To(Succeed())
+
+		var mgrContext context.Context
+		mgrContext, mgrCancel = context.WithCancel(ctx)
+
+		By("starting manager")
+		go func() {
+			err := mgr.Start(mgrContext)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+
+		// test client should be uncached and independent from the tested manager
+		c, err = client.New(cfg, client.Options{
+			Scheme: mgr.GetScheme(),
+			Mapper: mgr.GetRESTMapper(),
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(c).NotTo(BeNil())
+
+		flag.Parse()
+		validateFlags()
+
+		awsClient, err = awsclient.NewClient(*accessKeyID, *secretAccessKey, aws.DefaultDNSRegion)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterSuite(func() {
+		defer func() {
+			By("stopping manager")
+			mgrCancel()
+		}()
+
+		By("running cleanup actions")
+		framework.RunCleanupActions()
+
+		By("stopping test environment")
+		Expect(testEnv.Stop()).To(Succeed())
+	})
+
+	It("should successfully create and delete", func() {
+		By("setting up shoot environment")
+		Expect(c.Create(ctx, namespace)).To(Succeed())
+		Expect(c.Create(ctx, secret)).To(Succeed())
+		Expect(c.Create(ctx, cluster)).To(Succeed())
+		framework.AddCleanupAction(func() {
+			By("tearing down shoot environment")
+			Expect(client.IgnoreNotFound(c.Delete(ctx, cluster))).To(Succeed())
+			Expect(client.IgnoreNotFound(c.Delete(ctx, secret))).To(Succeed())
+			Expect(client.IgnoreNotFound(c.Delete(ctx, namespace))).To(Succeed())
+		})
+
+		By("creating dnsrecord")
+		dns := newDNSRecord(testName)
+		Expect(c.Create(ctx, dns)).To(Succeed())
+		framework.AddCleanupAction(func() {
+			By("deleting dnsrecord")
```

Comment:
Deletion logic should be executed in a `defer` block. Then in a separate cleanup action, the client should be used to clean up any orphaned resources. This avoid the need for manual cleanup, see https://github.com/gardener/terraformer/pull/57

Reply:
Yes, I had similar thoughts but didn't come to it yet. This is still not finished (so `Draft`). BTW, this is all copy pasted from the other integration tests (bastion in this case, but infra is the same), so ideally it should be fixed there as well.

Reply:
Yes, this should be fixed. Added a project card (https://github.com/orgs/gardener/projects/6#card-51424446) a long time ago, but didn't come to it yet.

Reply:
Done